### PR TITLE
feat: first cut gcloud/do integration

### PR DESCRIPTION
**Pull request by rsmitty**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -7,6 +7,7 @@ VPATH = $(PATH)
 KERNEL_IMAGE ?= autonomy/kernel:65ec2e6
 TOOLCHAIN_IMAGE ?= autonomy/toolchain:397b293
 GOLANG_VERSION ?= 1.11.4
+DOCKER_ARGS ?= ""
```
Comment (rsmitty) [inline Makefile]: This just adds the ability to pass mem/cpu limits to docker builds

File: Makefile [hunk 2]
```diff
@@ -112,7 +113,23 @@ installer:
 		--frontend-opt target=$@ \
 		$(COMMON_ARGS)
 	@docker load < build/$@.tar
-	@docker run --rm -v /dev:/dev -v $(PWD)/build:/out --privileged autonomy/$@:$(TAG) image -l
+
+do-image: installer
```
Comment (rsmitty) [inline Makefile]: three new make commands, one for each img type.

File: internal/pkg/userdata/userdata.go [hunk 3]
```diff
@@ -252,12 +252,21 @@ func (data *UserData) WriteFiles() (err error) {
 }
 
 // Download initializes a UserData struct from a remote URL.
-func Download(url string) (data *UserData, err error) {
+func Download(url string, headers *map[string]string) (data *UserData, err error) {
```
Comment (rsmitty) [inline internal/pkg/userdata/userdata.go]: Introduces a headers input for the download function. Gcloud requires you pass a "Metadata-Flavor": "Google"" header

File: internal/app/init/internal/platform/baremetal/baremetal.go [hunk 4]
```diff
@@ -86,7 +86,7 @@ func (b *BareMetal) UserData() (data *userdata.UserData, err error) {
 		return data, nil
 	}
 
-	return userdata.Download(option)
+	return userdata.Download(option, nil)
```
Comment (rsmitty) [inline internal/app/init/internal/platform/baremetal/baremetal.go]: Throwing nil vals in for headers on all the platform types that aren't gcloud. May be possible to do this better later on.

File: Makefile (hunk 2 above)
Comment (andrewrynhard) [reply Makefile]:
```suggestion
image-do: installer
```

File: internal/pkg/userdata/userdata.go (hunk 3 above)
Comment (andrewrynhard) [reply internal/pkg/userdata/userdata.go]: Can we add a TODO here?

```
TODO(andrewrynhard): Implement functional options.

File: Makefile [hunk 5]
```diff
@@ -112,7 +113,23 @@ installer:
...
+	-f -p digitalocean -u http://169.254.169.254/metadata/v1/user-data -e 'random.trust_cpu=on'
+
+gcloud-image: installer
+	@docker run --rm -v /dev:/dev -v $(PWD)/build/gcloud:/out --privileged $(DOCKER_ARGS) autonomy/installer:$(TAG) image -l \
+	-f -p googlecloud -u http://metadata.google.internal/computeMetadata/v1/instance/talos -e 'random.trust_cpu=on'
+	@mv $(PWD)/build/gcloud/image.raw $(PWD)/build/gcloud/disk.raw
+	##Hacks to get tar working properly
+	@export BUILDROOTDIR="$(PWD)"
+	@cd $(PWD)/build/gcloud
+	@tar -Sczf talos.tar.gz disk.raw
```
Comment (andrewrynhard) [inline Makefile]: We can remove the hack if we use `tar -C $(PWD)/build/gcloud -Sczf talos.tar.gz disk.raw`

File: Makefile [hunk 6]
```diff
@@ -112,7 +113,23 @@ installer:
 		--frontend-opt target=$@ \
 		$(COMMON_ARGS)
 	@docker load < build/$@.tar
-	@docker run --rm -v /dev:/dev -v $(PWD)/build:/out --privileged autonomy/$@:$(TAG) image -l
+
+do-image: installer
+	@docker run --rm -v /dev:/dev -v $(PWD)/build/do:/out --privileged $(DOCKER_ARGS) autonomy/installer:$(TAG) image -l \
+	-f -p digitalocean -u http://169.254.169.254/metadata/v1/user-data -e 'random.trust_cpu=on'
```
Comment (andrewrynhard) [inline Makefile]: The strategy I took with AWS was to ignore the `-u` flag and set it to `-u none` and then let each platform implementation use what they have defined as a constant.
```suggestion
	-f -p digitalocean -u none -e 'random.trust_cpu=on'
```

File: Makefile [hunk 7]
```diff
@@ -112,7 +113,23 @@ installer:
...
+
+gcloud-image: installer
+	@docker run --rm -v /dev:/dev -v $(PWD)/build/gcloud:/out --privileged $(DOCKER_ARGS) autonomy/installer:$(TAG) image -l \
+	-f -p googlecloud -u http://metadata.google.internal/computeMetadata/v1/instance/talos -e 'random.trust_cpu=on'
```
Comment (andrewrynhard) [inline Makefile]:
```suggestion
	-f -p googlecloud -u none -e 'random.trust_cpu=on'
```

File: Makefile (hunk 1 above)
Comment (andrewrynhard) [reply Makefile]: This was causing the docker run to fail... the docker command had a literal `""` that was throwing it off.
```suggestion
DOCKER_ARGS ?=
```

**Pull request closed without merging**